multicluster/service-mirror/cluster_watcher.go

```diff
@@ -207,7 +203,14 @@ func NewRemoteClusterServiceWatcher(
 		headlessServicesEnabled: enableHeadlessSvc,
 		// always instantiate the gatewayAlive=true to prevent unexpected service fail fast
 		gatewayAlive: true,
-	}, nil
+	}
+
+	_, err = remoteAPI.Client.Discovery().ServerVersion()
+	if err != nil {
+		return &rcsw, fmt.Errorf("cannot connect to api for target cluster %s: %w", clusterName, err)
```

Comment:
It's generally not considered sound to handle a value when err != nil; so we probably ought to omit the value.

```suggestion
		return nil, fmt.Errorf("cannot connect to api for target cluster %s: %w", clusterName, err)
```

However, your description seems to indicate that this is load-bearing:

> This fix reorders NewRemoteClusterServiceWatcher so that an object is returned even when there's an error, so cleanup on that object can be performed.

But the return value is not used at the call-site when an error is returned:

https://github.com/linkerd/linkerd2/blob/cf2999df42dd72802ce4eb60db6abdd9dc6efdca/multicluster/cmd/service-mirror/main.go#L306-L319

So, how does this change fix the problem exactly? How do we avoid introducing another problem like this. Can you add a comment so we don't easily run into this problem again?

Reply:
OH! this is setting a global value I don't think this is a sound pattern. Instead, the caller should use:

```go
 	 cw, err = servicemirror.NewRemoteClusterServiceWatcher( 
	 	ctx, 
	 	namespace, 
	 	controllerK8sAPI, 
	 	cfg, 
	 	&link, 
	 	requeueLimit, 
	 	repairPeriod, 
	 	ch, 
	 	enableHeadlessSvc, 
	 ) 
	 if err != nil { 
	 	return fmt.Errorf("unable to create cluster watcher: %w", err) 
	 }
     clusterWatcher = cw
```

This removes the need for the change to NewRemoteClusterServiceWatcher.

Reply:
If we don't modify NewRemoteClusterServiceWatcher to return `rcsw` on an error, the caller won't be able to perform the gauges cleanup. Actually I've just thought of something else; we should be able to perform the cleanup directly inside NewRemoteClusterServiceWatcher before returning the error. I've just pushed that, LMKWYT.

Reply:
I think this makes sense. Let's add a comment above this line explaining that the remoteAPI registers gauges and they must be explicitly unregistered on error. https://github.com/linkerd/linkerd2/pull/11875/files#diff-58391f2b0ac5849326792fbaf12a8e4aa8b06886acbe9fda308357d131ed38dcR172